libopenage/gui/game_control_link.h

```diff
@@ -0,0 +1,259 @@
+// Copyright 2015-2016 the openage authors. See copying.md for legal info.
+
+#pragma once
+
+#include <QQmlParserStatus>
+
+#include "../game_control.h"
+#include "guisys/link/gui_item.h"
+
+namespace openage {
+namespace gui {
+
+class GameControlLink;
+
+class OutputModeLink;
+
+template<>
+struct Wrap<OutputMode> {
+	typedef OutputModeLink Type;
+};
+
+template<>
+struct Unwrap<OutputModeLink> {
+	typedef OutputMode Type;
+};
+
+class OutputModeLink : public GuiItemQObject, public QQmlParserStatus, public GuiItemInterface<OutputModeLink> {
+Q_OBJECT
+
+Q_INTERFACES(QQmlParserStatus)
+Q_PROPERTY(QString name READ get_name NOTIFY name_changed)
+Q_PROPERTY(QStringList binds READ get_binds NOTIFY binds_changed)
+
+public:
+	OutputModeLink(QObject *parent=nullptr);
+	virtual ~OutputModeLink();
+
+	QString get_name() const;
+	QStringList get_binds() const;
+
+signals:
+	void name_changed();
+	void binds_changed();
+
+private slots:
+	void on_announced(const std::string &name, const std::vector<std::string>& binds);
+
+protected:
+	virtual void classBegin() override;
+	virtual void on_core_adopted() override;
+	virtual void componentComplete() override;
+
+private:
+	QString name;
+	QStringList binds;
+};
+
+class CreateModeLink;
+
+template<>
+struct Wrap<CreateMode> {
+	typedef CreateModeLink Type;
+};
+
+template<>
+struct Unwrap<CreateModeLink> {
+	typedef CreateMode Type;
+};
+
+class CreateModeLink : public Inherits<OutputModeLink, CreateModeLink> {
+Q_OBJECT
+
+public:
+	CreateModeLink(QObject *parent=nullptr);
+	virtual ~CreateModeLink();
+};
+
+class ActionModeLink;
+
+template<>
+struct Wrap<ActionMode> {
+	typedef ActionModeLink Type;
+};
+
+template<>
+struct Unwrap<ActionModeLink> {
+	typedef ActionMode Type;
+};
+
+class ActionModeLink : public Inherits<OutputModeLink, ActionModeLink> {
+Q_OBJECT
+
+// TODO: pack into a Q_GADGET in Qt 5.5
+Q_PROPERTY(int food READ get_food NOTIFY resources_changed)
+Q_PROPERTY(int wood READ get_wood NOTIFY resources_changed)
+Q_PROPERTY(int gold READ get_gold NOTIFY resources_changed)
+Q_PROPERTY(int stone READ get_stone NOTIFY resources_changed)
+Q_PROPERTY(QString ability READ get_ability NOTIFY ability_changed)
+
+public:
+	ActionModeLink(QObject *parent=nullptr);
+	virtual ~ActionModeLink();
+
+	int get_food() const;
+	int get_wood() const;
+	int get_gold() const;
+	int get_stone() const;
+	QString get_ability() const;
+
+signals:
+	void resources_changed();
+	void ability_changed();
+
+private slots:
+	void on_resources_changed(std::tuple<int, int, int, int> resources);
+	void on_ability_changed(const std::string &ability);
+
+private:
+	virtual void on_core_adopted() override;
+
+	int food;
+	int wood;
+	int gold;
+	int stone;
```

Comment:
the engine should especially be capable running star wars galactic battlegrounds some day as it's the exact same engine just with another game loaded.